src/Lib.hs

```diff
@@ -59,3 +59,16 @@ checkShadowing args (Abs arg expr)
       nested  = checkShadowing (arg : args) expr
 checkShadowing args (App t u) = concatMap (checkShadowing args) [t, u]
 checkShadowing _ _ = []
+
+etaReduce :: Expr -> Expr
+etaReduce expr = case expr of
+  (Abs ident (App expr' (Term ident'))) ->
+    if ident == ident' && isNotFreeVar then expr' else expr
+      where isNotFreeVar = not $ isFreeVarOf ident expr'
+  _                                     -> expr
+
+isFreeVarOf :: Identifier -> Expr -> Bool
+isFreeVarOf var (Term t)         = var == t
+isFreeVarOf var (Abs ident body) = var /= ident && isFreeVarOf var body
+isFreeVarOf var (App t u)        = isFreeVarOf var t || isFreeVarOf var u
+isFreeVarOf _ _                  = True
```

Comment:
Let's write a function that checks whether an `Expr` is a valid lambda calculus "statement" using rules such as above. This function can return `Maybe Expr` for now and then later extended to `Either` (and so on) and a newtype wrapper (e.g. `Valid Expr`) so that functions like etaReduce would only work with checked expressions. I'll create an issue for that I think